compiler/infra/ILTraverser.hpp

```diff
@@ -0,0 +1,173 @@
+ ******************************************************************************/
+
+#ifndef TR_ILTRAVERSER_HPP
+#define TR_ILTRAVERSER_HPP
+
+#include "infra/Checklist.hpp"
+#include "env/TRMemory.hpp"
+#include "compile/Compilation.hpp"
+
+#include <vector>
+#include <algorithm>
+
+namespace TR {
+
+// forward declaration of "imported" classes
+class ResolvedMethodSymbol;
+class Block;
+class TreeTop;
+class Node;
+
+/**
+ * @brief A class that abstracts aways the details of traversing Testarossa IL.
+ *
+ * This class uses the Observer pattern to separate the actual traversal mechanism
+ * from the code that needs to do the traversal. An observer can register itself
+ * with an instance of this class to be notified of the various transitions during
+ * the traversal.
+ *
+ * This class expects the input IL to be "sound"; properly structure for traversal.
+ */
+class ILTraverser
+   {
+   public:
+   explicit ILTraverser(TR::Compilation* comp);
+
+   /**
+    * @brief The base class for observers of an instance of ILTraverser
+    *
+    * Objects inheriting from this class can be registered with an instance of
+    * ILTraverser to be notified of the various transitions done during traversal.
+    *
+    * All signals have an empty body by default so that derived classes need
+    * only implement the signals they are interested in.
+    *
+    * All the events that can be listened to have the same general interface.
+    * They all take as argument a pointer to the object being visited. For each
+    * type of object that can be visited, there is one singal for when the object
+    * is entered, and one signal for when traversal returns to the object after
+    * having visited the object's children.
+    *
+    *    parent      --,
+    *       node    <--'  visitingNode(node)
+    *          child
+    *
+    *    parent
+    *       node        <--,  returnedToNode(node)
+    *          child     --'
+    *
+    * Some specialized versions of these signals are also provided to accomodate
+    * the traversal requirements of certain object types (e.g. there is a
+    * special signal for when a commoned node is being re-visited).
+    */
+   struct Observer
+      {
+      virtual ~Observer() = default;
+
+      // emitted when visiting methods
+      virtual void visitingMethod(TR::ResolvedMethodSymbol* method) {}
+      virtual void returnedToMethod(TR::ResolvedMethodSymbol* method) {}
+
+      // emitted when visiting blocks
+      virtual void visitingBlock(TR::Block* block) {}
+      virtual void returnedToBlock(TR::Block* block) {}
+
+      // emitted when visiting TreeTops
+      virtual void visitingTreeTop(TR::TreeTop* treeTop) {}
+      virtual void returnedToTreeTop(TR::TreeTop* treeTop) {}
+
+      // emitted when visiting IL nodes
+      virtual void visitingNode(TR::Node* node) {}
+      virtual void visitingCommonedChildNode(TR::Node* node) {}
+      virtual void returningToNode(TR::Node* node) {}
```

Comment:
Shouldn't this be `returnedToNode`?  That's how its documented and how it appears to be implemented.